### PR TITLE
[CI] Don't do coverage builds for engineless fuzzers. 

### PR DESCRIPTION
**Pull request by jonathanmetzman**
Coverage builds need to be special cased since they aren't specified
in sanitizers. Instead they are done for all C/C++ projects that
use libFuzzer.

Move all of this special casing to `should_build_coverage` and call that from
`should_build` so we have one place where we decide this.
Add tests as well.

Fixes: https://github.com/google/oss-fuzz/issues/4371

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
Please ignore the last 6 commits, I was doing them to confirm that CI works properly now in the general case and for engineless projects.

**Comment by inferno-chromium** [comment]
Sorry i wont have time for review this week, if @oliverchang has time can take a look, or defer to next week.

File: infra/ci/build.py [hunk 1]
```diff
@@ -87,8 +87,34 @@ def check_build(project, engine, sanitizer, architecture):
...
 
+def should_build_coverage(project_yaml):
+  """Returns True if a coverage build should be done based on project.yaml
+  contents."""
+  # Enable coverage builds on projects that use engines. Those that don't use
+  # engines shouldn't get coverage builds.
+  engines = project_yaml.get('fuzzing_engines', DEFAULT_ENGINES)
+  engineless = 'none' in engines
+  if engineless:
+    assert_message = 'Project supports engines and none: {engines}.'.format(
```
Comment (oliverchang) [inline infra/ci/build.py:98]: This message seems a bit confusing.

Consider rewording along the lines of `'"none" engine must be the only one specified.'`

File: infra/ci/build.py [hunk 2]
```diff
@@ -87,8 +87,34 @@ def check_build(project, engine, sanitizer, architecture):
...
+        engines=engines)
+    assert len(engines) == 1, assert_message
+    return False
+
+  language = project_yaml.get('language')
+  if language not in LANGUAGES_WITH_COVERAGE_SUPPORT:
+    print(('Project is written in "{language}", '
```
Comment (oliverchang) [inline infra/ci/build.py:105]: Probably for another PR, but if everything is on Python 3 now, we can use f-strings to reduce a lot of the clutter with .format() calls.

**Review by oliverchang — APPROVED** [review]

File: infra/ci/build.py (hunk 2 above)
Comment (jonathanmetzman) [reply infra/ci/build.py:105]: Good point although I don't really love the idea of f-strings. They're probably fine for just variable expansion but I think anything more than that like expressions could start causing problems.
I think we could reduce a lot of this clutter if we didn't insist on using `{language}` and `.format(language=blah)` I don't  push back on this when I'm told do so but I think doing this for a string with a single variable just causes clutter without really improving readability. Especially for log lines where it isn't catastrophic if things are wrong.
Instead we could use `{}`, or `{0}` or even the printf style (`'%s' % blah`).
I'll probably lose this argument so f-strings will be a nice way to keep the explicitness without hurting readability.

Comment (jonathanmetzman) [reply infra/ci/build.py:105]: But yeah I think this should come in another PR.

Comment (jonathanmetzman) [reply infra/ci/build.py:105]: FWIW the style guide doesn't complain about f-strings so maybe I'm being overly skeptical.

File: infra/ci/build.py (hunk 1 above)
Comment (jonathanmetzman) [reply infra/ci/build.py:98]: Done.

File: infra/ci/build.py (hunk 2 above)
Comment (jonathanmetzman) [reply infra/ci/build.py:105]: Another thing we should consider with printing errors is that github actions usually writes all of stderr and then stdout. So printing can cause stuff to look out of order. Maybe we can change things so everything is written to stderr.

**Pull request merged**